python_bindings/src/builtin_helpers_src.py

```diff
@@ -0,0 +1,787 @@
+
+    def _make_replacement(self, value: Any, r: Requirement) -> Func:
+        assert _is_valid_name(r._name) and len(r._types) > 0 and r._dims >= 0
+        if value is None:
+            return Func(r._types, r._dims, r._name)
+
+        _check(
+            isinstance(value, (Buffer, Func)),
+            "Output %s requires a Func or Buffer argument when using call(), but saw %s"
+            % (r._name, str(value)),
+        )
+        _check(
+            value.defined(),
+            "Cannot set the value for %s to an undefined value." % r._name,
+        )
+
+        if isinstance(value, Buffer):
+            r._check_types_and_dims([value.type()], value.dimensions())
+        else:
+            r._check_types_and_dims(value.output_types(), value.dimensions())
+
+        if isinstance(value, Buffer):
+            # Allow assignment from a Buffer<> to an Output<Buffer<>>;
+            # this allows us to use a statically-compiled buffer inside a Generator
+            # to assign to an output.
+            f = Func(r._types, r._dims, r._name)
+            f[_] = value[_]
+            value = f
+
+        return value
+
+
+# OutputScalar is just like OutputBuffer, except it is always dimensions = 0
+class OutputScalar(OutputBuffer):
+
+    def __init__(self, types: Optional[Type]):
+        OutputBuffer.__init__(self, types, 0)
+
+    def _make_replacement(self, value: Any, r: Requirement) -> Func:
+        assert _is_valid_name(r._name) and len(r._types) > 0 and r._dims == 0
+        if value is None:
+            return Func(r._types, 0, r._name)
+
+        _check(
+            isinstance(value, Expr),
+            "Output %s requires an Expr argument when using call(), but saw %s"
+            % (r._name, str(value)),
+        )
+        _check(
+            value.defined(),
+            "Cannot set the value for %s to an undefined value." % r._name,
+        )
+        r._check_types_and_dims(value.type(), 0)
+        f = Func(value.type(), 0, r._name)
+        f[_] = value
+        return f
+
+
+@total_ordering
+class _Stage(Enum):
+    Created = 0
+    GeneratorParamsBuilt = 1
+    IOBuilt = 2
+    IOFinalized = 3
+    PipelineBuilt = 4
+
+    def __lt__(self, other):
+        if self.__class__ is other.__class__:
+            return self.value < other.value
+        return NotImplemented
+
+
+# Prevent user code from inadvertently overwriting fields -- this is
+# not at all foolproof but will make it inconvenient to do so.
+# TODO: add ability to freeze ImageParam and Param so you can't call set(), etc.
+def _make_freezable(cls: type):
+
+    # TODO: kinda janky, is this the best way to implement this?
+    def _freeze(self):
+        setattr(self, "_frozen", True)
+
+    def _freezable_setattr(self, name, value):
+        if getattr(self, "_frozen", False):
+            raise AttributeError("Invalid write to field '%s'" % name)
+        self.__class__._old_halide_setattr(self, name, value)
```

Comment:
Are we concerned about users setting their own `__setattr__` hooks?

Reply:
Uh, to be honest... I'm not sure. I just assumed that was the appropriate thing to do, since we're injecting a change into a possibly-arbitrary class. Is it more common to just overwrite?

Reply:
It's fine, it just looked defensive. I would have just overwritten and used `object.__setattr__`.